File: python/geocercas.py
"""
==========================================================================
GEOCERCAS.PY - Algoritmo de Geocercas Dinámicas en Python
Clave del Peluquero Exitoso
Fecha: 2026-04-06
Descripción: Implementación del algoritmo de agrupamiento logístico
             basado en proximidad euclidiana y capacidad máxima.
             Diseñado para ejecutarse como módulo backend o script standalone.

Uso:
    python geocercas.py --pedidos 100 --dmax 0.015 --capmax 15

Objetivo Específico:
    Modelar el algoritmo de geocercas dinámicas utilizando umbrales de
    proximidad euclidiana y capacidad de carga máxima, verificando la
    condición de término del ciclo principal y la correcta inicialización
    de variables acumuladoras para garantizar la ausencia de errores
    lógicos (ciclos infinitos o condiciones inalcanzables).
==========================================================================
"""

import math
import time
import random
import sys
import json
from dataclasses import dataclass, field
from typing import List, Tuple, Optional

# ============================================================
# 1. CONSTANTES DE CONFIGURACIÓN
# ============================================================

# Centro de Caldas, Antioquia (Carrera 49 #134 sur-41)
CENTRO_LAT: float = 6.090722
CENTRO_LNG: float = -75.638787

# Radio de dispersión para generar pedidos (~2km en grados)
RADIO_DISPERSION: float = 0.018

# Productos disponibles para simulación
PRODUCTOS: List[str] = [
    "Shampoo Profesional Keratina 500ml",
    "Kit Tinte + Oxidante 90ml",
    "Proteína Capilar Reconstructora 250g",
    "Secador Profesional 2200W",
    "Plancha de Titanio Pro",
    "Acondicionador Argán 400ml",
    "Cepillo Cerámico Iónico",
    "Suero Capilar Reparador 50ml",
    "Tijeras Profesionales 6.5 pulgadas",
    "Spray Fijador Extra Fuerte 300ml",
]


# ============================================================
# 2. ESTRUCTURA DE DATOS: Pedido
# ============================================================

@dataclass
class Pedido:
    """
    Representa un pedido/paquete con ubicación geográfica.

    Attributes:
        id (int): Identificador único del pedido
        lat (float): Latitud del destino de entrega
        lng (float): Longitud del destino de entrega
        peso (float): Peso del paquete en kilogramos
        producto (str): Nombre del producto
        codigo (str): Código de rastreo único (CLV-2026-XXXXX)
        grupo (int): Índice de la geocerca asignada (-1 = sin asignar)
    """
    id: int
    lat: float
    lng: float
    peso: float
    producto: str = ""
    codigo: str = ""
    grupo: int = -1  # -1 indica que no está asignado a ninguna geocerca


@dataclass
class Centroide:
    """
    Representa el centro geométrico de una geocerca.

    Attributes:
        lat (float): Latitud del centroide
        lng (float): Longitud del centroide
    """
    lat: float
    lng: float


# ============================================================
# 3. GENERACIÓN DE DATOS SIMULADOS
# ============================================================

def generar_codigo_rastreo() -> str:
    """
    Genera un código de rastreo único tipo 'CLV-2026-XXXXX'.

    Returns:
        str: Código de rastreo alfanumérico
    """
    num = random.randint(10000, 99999)
    return f"CLV-2026-{num}"


def generar_pedidos(n: int) -> List[Pedido]:
    """
    Genera N pedidos aleatorios alrededor de Caldas, Antioquia.
    Utiliza distribución gaussiana (Box-Muller) para una dispersión
    más realista alrededor del centro.

    Args:
        n (int): Cantidad de pedidos a generar

    Returns:
        List[Pedido]: Lista de pedidos generados con coordenadas
    """
    pedidos: List[Pedido] = []

    for i in range(n):
        # -- Distribución gaussiana (Box-Muller) --
        u1 = random.random()
        u2 = random.random()
        # Evitar log(0) que causaría error matemático
        if u1 == 0:
            u1 = 1e-10
        z0 = math.sqrt(-2 * math.log(u1)) * math.cos(2 * math.pi * u2)
        z1 = math.sqrt(-2 * math.log(u1)) * math.sin(2 * math.pi * u2)

        lat = CENTRO_LAT + z0 * RADIO_DISPERSION
        lng = CENTRO_LNG + z1 * RADIO_DISPERSION
        peso = round(0.2 + random.random() * 4.8, 2)  # 0.2 a 5.0 kg
        producto = random.choice(PRODUCTOS)
        codigo = generar_codigo_rastreo()

        pedidos.append(Pedido(
            id=i + 1,
            lat=lat,
            lng=lng,
            peso=peso,
            producto=producto,
            codigo=codigo
        ))

    return pedidos


# ============================================================
# 4. ALGORITMO DE AGRUPAMIENTO POR GEOCERCAS
# ============================================================

def distancia_euclidiana(lat1: float, lng1: float,
                         lat2: float, lng2: float) -> float:
    """
    Calcula la distancia euclidiana entre dos puntos geográficos.
    Se usa la fórmula simplificada para distancias cortas (mismo municipio).

    Args:
        lat1, lng1: Coordenadas del punto 1
        lat2, lng2: Coordenadas del punto 2

    Returns:
        float: Distancia euclidiana en grados

    Nota:
        Para distancias largas se debería usar la fórmula de Haversine,
        pero para un municipio (~5km) la diferencia es despreciable.
    """
    d_lat = lat2 - lat1
    d_lng = lng2 - lng1
    return math.sqrt(d_lat ** 2 + d_lng ** 2)


def agrupar_geocercas(pedidos: List[Pedido],
                       d_max: float,
                       cap_max: int) -> Tuple[List[List[Pedido]], List[Centroide]]:
    """
    Algoritmo principal de agrupamiento por geocercas dinámicas.

    LÓGICA DEL ALGORITMO:
    ─────────────────────
    1. Inicializar listas de grupos y centroides (vacías)
    2. Para cada pedido (iteración sobre N pedidos):
       a. Buscar entre todos los grupos existentes el más cercano
          cuyo centroide esté a distancia ≤ d_max
       b. Verificar que el grupo no exceda cap_max pedidos
       c. Si se encuentra grupo válido → asignar pedido y
          recalcular centroide con media incremental
       d. Si no se encuentra → crear nuevo grupo (semilla)
    3. Retornar listas de grupos y centroides

    VERIFICACIÓN DE ESTRUCTURA DE CONTROL:
    ──────────────────────────────────────
    - Ciclo externo: for i in range(len(pedidos))
      → Itera exactamente N veces, SIEMPRE termina
      → No es un while, por lo tanto NO hay riesgo de ciclo infinito

    - Ciclo interno: for g in range(len(grupos))
      → Itera sobre grupos existentes (máximo N grupos)
      → Se ejecuta como máximo N×N = N² veces en total

    - Variables acumuladoras:
      → grupos: List[List[Pedido]] = [] (inicializada vacía ✓)
      → centroides: List[Centroide] = [] (inicializada vacía ✓)
      → pesos_grupo: List[float] = [] (inicializada vacía ✓)
      → mejor_grupo: int = -1 (reinicializada en cada iteración ✓)
      → mejor_distancia: float = inf (reinicializada en cada iteración ✓)

    COMPLEJIDAD:
    → Tiempo: O(n × k) donde n = pedidos, k = grupos formados
    → Espacio: O(n) para almacenar las asignaciones

    Args:
        pedidos (List[Pedido]): Lista de pedidos a agrupar
        d_max (float): Distancia máxima (umbral de proximidad) en grados
        cap_max (int): Capacidad máxima de paquetes por geocerca

    Returns:
        Tuple: (grupos: List[List[Pedido]], centroides: List[Centroide])
    """
    # -- Inicialización de variables acumuladoras --
    grupos: List[List[Pedido]] = []       # Cada elemento es un grupo de pedidos
    centroides: List[Centroide] = []       # Centroide de cada grupo
    pesos_grupo: List[float] = []          # Peso acumulado por grupo

    # -- Ciclo principal: iterar exactamente N veces (siempre termina) --
    for i in range(len(pedidos)):
        pedido = pedidos[i]

        # Variables de búsqueda reinicializadas en cada iteración
        mejor_grupo: int = -1               # Índice del grupo más cercano
        mejor_distancia: float = float('inf')  # Distancia mínima encontrada

        # -- Buscar grupo más cercano que cumpla restricciones --
        for g in range(len(grupos)):
            # Restricción 1: No exceder capacidad máxima
            if len(grupos[g]) >= cap_max:
                continue

            # Calcular distancia euclidiana al centroide
            dist = distancia_euclidiana(
                pedido.lat, pedido.lng,
                centroides[g].lat, centroides[g].lng
            )

            # Restricción 2: Debe estar dentro del radio máximo
            if dist <= d_max and dist < mejor_distancia:
                mejor_grupo = g
                mejor_distancia = dist

        # -- Decisión: asignar a grupo existente o crear nuevo --
        if mejor_grupo != -1:
            # Asignar al grupo más cercano
            grupos[mejor_grupo].append(pedido)
            pesos_grupo[mejor_grupo] += pedido.peso
            pedido.grupo = mejor_grupo

            # Recalcular centroide (media incremental)
            n = len(grupos[mejor_grupo])
            centroides[mejor_grupo].lat = (
                (centroides[mejor_grupo].lat * (n - 1) + pedido.lat) / n
            )
            centroides[mejor_grupo].lng = (
                (centroides[mejor_grupo].lng * (n - 1) + pedido.lng) / n
            )
        else:
            # Crear nuevo grupo con este pedido como semilla
            pedido.grupo = len(grupos)
            grupos.append([pedido])
            centroides.append(Centroide(lat=pedido.lat, lng=pedido.lng))
            pesos_grupo.append(pedido.peso)

    return grupos, centroides


# ============================================================
# 5. MEDICIÓN DE RENDIMIENTO (BENCHMARK)
# ============================================================

def ejecutar_benchmark(pedidos: List[Pedido],
                        d_max: float,
                        cap_max: int) -> dict:
    """
    Ejecuta el algoritmo midiendo tiempo de ejecución y uso de memoria.

    Args:
        pedidos: Lista de pedidos a procesar
        d_max: Distancia máxima en grados
        cap_max: Capacidad máxima por geocerca

    Returns:
        dict: Resultados con métricas de rendimiento
    """
    # -- Medir tiempo de ejecución --
    inicio = time.perf_counter()

    # -- Medir memoria (aproximado) --
    import tracemalloc
    tracemalloc.start()

    # Ejecutar algoritmo
    grupos, centroides = agrupar_geocercas(pedidos, d_max, cap_max)

    # Obtener métricas
    _, pico_memoria = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    fin = time.perf_counter()

    tiempo_ms = (fin - inicio) * 1000  # Convertir a milisegundos
    memoria_mb = pico_memoria / (1024 * 1024)  # Convertir a megabytes

    return {
        "lenguaje": "Python",
        "num_pedidos": len(pedidos),
        "num_grupos": len(grupos),
        "tiempo_ms": round(tiempo_ms, 3),
        "memoria_mb": round(memoria_mb, 3),
        "grupos": [
            {
                "id": i,
                "num_pedidos": len(g),
                "centroide": {"lat": c.lat, "lng": c.lng},
                "peso_total": round(sum(p.peso for p in g), 2)
            }
            for i, (g, c) in enumerate(zip(grupos, centroides))
        ]
    }


# ============================================================
# 6. INTERFAZ DE LÍNEA DE COMANDOS
# ============================================================

def main():
    """
    Punto de entrada principal para ejecución desde terminal.

    Uso:
        python geocercas.py --pedidos 100 --dmax 0.015 --capmax 15
        python geocercas.py  (usa valores por defecto)
    """
    import argparse

    parser = argparse.ArgumentParser(
        description="Algoritmo de Geocercas Dinámicas - Clave del Peluquero"
    )
    parser.add_argument("--pedidos", type=int, default=100,
                        help="Número de pedidos a generar (default: 100)")
    parser.add_argument("--dmax", type=float, default=0.015,
                        help="Distancia máxima en grados (default: 0.015)")
    parser.add_argument("--capmax", type=int, default=15,
                        help="Capacidad máxima por geocerca (default: 15)")
    parser.add_argument("--json", action="store_true",
                        help="Salida en formato JSON")

    args = parser.parse_args()

    print(f"\n{'='*60}")
    print(f"  GEOCERCAS DINÁMICAS - Clave del Peluquero Exitoso")
    print(f"  Centro: Caldas, Antioquia ({CENTRO_LAT}, {CENTRO_LNG})")
    print(f"{'='*60}\n")

    # Generar pedidos
    print(f"📦 Generando {args.pedidos} pedidos aleatorios...")
    pedidos = generar_pedidos(args.pedidos)

    # Ejecutar benchmark
    print(f"🔄 Ejecutando algoritmo (d_max={args.dmax}, cap_max={args.capmax})...")
    resultado = ejecutar_benchmark(pedidos, args.dmax, args.capmax)

    if args.json:
        # Salida JSON para integración con frontend
        print(json.dumps(resultado, indent=2, ensure_ascii=False))
    else:
        # Salida legible
        print(f"\n{'─'*40}")
        print(f"  RESULTADOS")
        print(f"{'─'*40}")
        print(f"  Lenguaje:        {resultado['lenguaje']}")
        print(f"  Pedidos:         {resultado['num_pedidos']}")
        print(f"  Geocercas:       {resultado['num_grupos']}")
        print(f"  Tiempo:          {resultado['tiempo_ms']:.3f} ms")
        print(f"  Memoria pico:    {resultado['memoria_mb']:.3f} MB")
        print(f"{'─'*40}\n")

        # Detalle por geocerca
        for g in resultado["grupos"]:
            print(f"  📍 Geocerca #{g['id']+1}: "
                  f"{g['num_pedidos']} pedidos, "
                  f"{g['peso_total']} kg, "
                  f"centro ({g['centroide']['lat']:.4f}, "
                  f"{g['centroide']['lng']:.4f})")

    print(f"\n✅ Algoritmo completado exitosamente.\n")


if __name__ == "__main__":
    main()

File: cpp/geocercas.cpp
/**
 * ==========================================================================
 * GEOCERCAS.CPP - Algoritmo de Geocercas Dinámicas en C++
 * Clave del Peluquero Exitoso
 * Fecha: 2026-04-06
 * Descripción: Implementación optimizada del algoritmo de agrupamiento
 *              logístico para compilar a WebAssembly (WASM) con Emscripten.
 *
 * Compilación nativa:
 *   g++ -O2 -std=c++17 -o geocercas geocercas.cpp
 *   ./geocercas --pedidos 100 --dmax 0.015 --capmax 15
 *
 * Compilación a WASM (requiere Emscripten SDK):
 *   emcc -O2 -std=c++17 -s WASM=1 -s EXPORTED_FUNCTIONS="['_agruparGeocercas']"
 *        -o geocercas.js geocercas.cpp
 *
 * Objetivo Específico:
 *   Implementar el algoritmo de geocercas dinámicas en C++ utilizando
 *   estructuras STL (vector, cmath) con medición de rendimiento mediante
 *   <chrono>, verificando la correcta inicialización de variables
 *   acumuladoras y la condición de término del ciclo principal.
 * ==========================================================================
 */

#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <random>
#include <string>
#include <iomanip>
#include <limits>
#include <sstream>

// ============================================================
// 1. CONSTANTES DE CONFIGURACIÓN
// ============================================================

/** Centro de Caldas, Antioquia (Carrera 49 #134 sur-41) */
constexpr double CENTRO_LAT = 6.090722;
constexpr double CENTRO_LNG = -75.638787;

/** Radio de dispersión para generar pedidos (~2km en grados) */
constexpr double RADIO_DISPERSION = 0.018;

/** Pi para cálculos trigonométricos */
constexpr double PI = 3.14159265358979323846;


// ============================================================
// 2. ESTRUCTURAS DE DATOS
// ============================================================

/**
 * @struct Pedido
 * @brief Representa un pedido/paquete con ubicación geográfica.
 *
 * Cada pedido tiene coordenadas de destino, peso del paquete,
 * nombre del producto y un índice de grupo asignado.
 */
struct Pedido {
    int id;              ///< Identificador único del pedido
    double lat;          ///< Latitud del destino de entrega
    double lng;          ///< Longitud del destino de entrega
    double peso;         ///< Peso del paquete en kilogramos
    std::string producto; ///< Nombre del producto
    std::string codigo;  ///< Código de rastreo (CLV-2026-XXXXX)
    int grupo;           ///< Geocerca asignada (-1 = sin asignar)

    /// Constructor por defecto: inicializa grupo a -1 (sin asignar)
    Pedido() : id(0), lat(0), lng(0), peso(0), grupo(-1) {}

    /// Constructor con parámetros
    Pedido(int _id, double _lat, double _lng, double _peso,
           const std::string& _producto, const std::string& _codigo)
        : id(_id), lat(_lat), lng(_lng), peso(_peso),
          producto(_producto), codigo(_codigo), grupo(-1) {}
};


/**
 * @struct Centroide
 * @brief Centro geométrico de una geocerca.
 */
struct Centroide {
    double lat;  ///< Latitud del centroide
    double lng;  ///< Longitud del centroide

    Centroide(double _lat = 0, double _lng = 0) : lat(_lat), lng(_lng) {}
};


/**
 * @struct ResultadoBenchmark
 * @brief Métricas de rendimiento del algoritmo.
 */
struct ResultadoBenchmark {
    int num_pedidos;     ///< Cantidad de pedidos procesados
    int num_grupos;      ///< Cantidad de geocercas formadas
    double tiempo_ms;    ///< Tiempo de ejecución en milisegundos
    double memoria_mb;   ///< Memoria utilizada (estimada) en megabytes
};


// ============================================================
// 3. GENERACIÓN DE DATOS SIMULADOS
// ============================================================

/**
 * @brief Genera un código de rastreo único tipo "CLV-2026-XXXXX"
 * @param rng Generador de números aleatorios
 * @return Código de rastreo como string
 */
std::string generarCodigoRastreo(std::mt19937& rng) {
    std::uniform_int_distribution<int> dist(10000, 99999);
    return "CLV-2026-" + std::to_string(dist(rng));
}


/**
 * @brief Genera N pedidos aleatorios alrededor de Caldas, Antioquia.
 *
 * Utiliza distribución gaussiana (Box-Muller) para dispersión realista.
 *
 * @param n Cantidad de pedidos a generar
 * @return Vector de pedidos con coordenadas aleatorias
 */
std::vector<Pedido> generarPedidos(int n) {
    // -- Inicialización del generador aleatorio con semilla --
    std::mt19937 rng(std::chrono::steady_clock::now().time_since_epoch().count());
    std::uniform_real_distribution<double> uniforme(0.0, 1.0);
    std::uniform_real_distribution<double> peso_dist(0.2, 5.0);

    // -- Lista de productos para simulación --
    const std::vector<std::string> productos = {
        "Shampoo Profesional Keratina 500ml",
        "Kit Tinte + Oxidante 90ml",
        "Proteina Capilar Reconstructora 250g",
        "Secador Profesional 2200W",
        "Plancha de Titanio Pro",
        "Acondicionador Argan 400ml",
        "Cepillo Ceramico Ionico",
        "Suero Capilar Reparador 50ml",
        "Tijeras Profesionales 6.5 pulgadas",
        "Spray Fijador Extra Fuerte 300ml"
    };

    std::vector<Pedido> pedidos;
    pedidos.reserve(n);  // Reservar memoria de antemano

    for (int i = 0; i < n; ++i) {
        // -- Distribución gaussiana con Box-Muller --
        double u1 = uniforme(rng);
        double u2 = uniforme(rng);
        if (u1 < 1e-10) u1 = 1e-10;  // Evitar log(0)

        double z0 = std::sqrt(-2.0 * std::log(u1)) * std::cos(2.0 * PI * u2);
        double z1 = std::sqrt(-2.0 * std::log(u1)) * std::sin(2.0 * PI * u2);

        double lat = CENTRO_LAT + z0 * RADIO_DISPERSION;
        double lng = CENTRO_LNG + z1 * RADIO_DISPERSION;
        double peso = std::round(peso_dist(rng) * 100.0) / 100.0;

        // Seleccionar producto aleatorio
        std::uniform_int_distribution<int> prod_dist(0, productos.size() - 1);
        std::string producto = productos[prod_dist(rng)];
        std::string codigo = generarCodigoRastreo(rng);

        pedidos.emplace_back(i + 1, lat, lng, peso, producto, codigo);
    }

    return pedidos;
}


// ============================================================
// 4. ALGORITMO DE AGRUPAMIENTO POR GEOCERCAS
// ============================================================

/**
 * @brief Calcula la distancia euclidiana entre dos puntos geográficos.
 *
 * Fórmula simplificada para distancias cortas (mismo municipio).
 * Para distancias largas se debería usar Haversine.
 *
 * @param lat1, lng1 Coordenadas del punto 1
 * @param lat2, lng2 Coordenadas del punto 2
 * @return Distancia euclidiana en grados
 */
inline double distanciaEuclidiana(double lat1, double lng1,
                                   double lat2, double lng2) {
    double dLat = lat2 - lat1;
    double dLng = lng2 - lng1;
    return std::sqrt(dLat * dLat + dLng * dLng);
}


/**
 * @brief Algoritmo principal de agrupamiento por geocercas dinámicas.
 *
 * LÓGICA:
 *   1. Inicializar vectores de grupos y centroides (vacíos)
 *   2. Para cada pedido (i = 0 hasta n-1):
 *      a. Buscar grupo existente más cercano con dist ≤ dMax
 *      b. Verificar capacidad (grupo.size() < capMax)
 *      c. Si existe → asignar y recalcular centroide
 *      d. Si no → crear nuevo grupo con pedido como semilla
 *   3. Retornar estructuras de datos
 *
 * VERIFICACIÓN DE TERMINACIÓN:
 *   - Ciclo externo: for(i=0; i<n; i++) → siempre termina en N pasos
 *   - Ciclo interno: for(g=0; g<grupos.size(); g++) → termina en K pasos
 *   - No hay condiciones inalcanzables
 *   - Variables acumuladoras correctamente inicializadas
 *
 * COMPLEJIDAD: O(n × k) donde n = pedidos, k = grupos
 *
 * @param pedidos Vector de pedidos a agrupar
 * @param dMax Distancia máxima (umbral de proximidad) en grados
 * @param capMax Capacidad máxima de paquetes por geocerca
 * @param[out] out_centroides Vector de centroides de salida
 * @return Vector de vectores (cada sub-vector es un grupo)
 */
std::vector<std::vector<Pedido>> agruparGeocercas(
    std::vector<Pedido>& pedidos,
    double dMax,
    int capMax,
    std::vector<Centroide>& out_centroides)
{
    // -- Variables acumuladoras inicializadas correctamente --
    std::vector<std::vector<Pedido>> grupos;    // Grupos formados (vacío ✓)
    std::vector<Centroide> centroides;           // Centroides (vacío ✓)
    std::vector<double> pesos_grupo;             // Pesos acumulados (vacío ✓)

    int n = static_cast<int>(pedidos.size());

    // -- Ciclo principal: iterar exactamente N veces --
    // Condición de término: i < n (siempre alcanzable, no es ciclo infinito)
    for (int i = 0; i < n; ++i) {
        Pedido& pedido = pedidos[i];

        // Variables de búsqueda reinicializadas en cada iteración
        int mejorGrupo = -1;                    // Índice del mejor grupo
        double mejorDistancia = std::numeric_limits<double>::infinity();

        // -- Buscar grupo más cercano (ciclo interno) --
        int numGrupos = static_cast<int>(grupos.size());
        for (int g = 0; g < numGrupos; ++g) {
            // Restricción 1: Capacidad máxima
            if (static_cast<int>(grupos[g].size()) >= capMax) {
                continue;
            }

            // Calcular distancia euclidiana al centroide
            double dist = distanciaEuclidiana(
                pedido.lat, pedido.lng,
                centroides[g].lat, centroides[g].lng
            );

            // Restricción 2: Radio máximo
            if (dist <= dMax && dist < mejorDistancia) {
                mejorGrupo = g;
                mejorDistancia = dist;
            }
        }

        // -- Decisión de asignación --
        if (mejorGrupo != -1) {
            // Asignar al grupo existente más cercano
            grupos[mejorGrupo].push_back(pedido);
            pesos_grupo[mejorGrupo] += pedido.peso;
            pedido.grupo = mejorGrupo;

            // Recalcular centroide (media incremental)
            int m = static_cast<int>(grupos[mejorGrupo].size());
            centroides[mejorGrupo].lat =
                ((centroides[mejorGrupo].lat * (m - 1)) + pedido.lat) / m;
            centroides[mejorGrupo].lng =
                ((centroides[mejorGrupo].lng * (m - 1)) + pedido.lng) / m;
        } else {
            // Crear nuevo grupo con este pedido como semilla
            pedido.grupo = static_cast<int>(grupos.size());
            grupos.push_back({pedido});
            centroides.emplace_back(pedido.lat, pedido.lng);
            pesos_grupo.push_back(pedido.peso);
        }
    }

    // Copiar centroides al parámetro de salida
    out_centroides = centroides;

    return grupos;
}


// ============================================================
// 5. FUNCIÓN PRINCIPAL Y BENCHMARK
// ============================================================

/**
 * @brief Punto de entrada principal. Ejecuta el benchmark completo.
 *
 * Uso:
 *   ./geocercas                               (valores por defecto)
 *   ./geocercas --pedidos 500 --dmax 0.02      (personalizado)
 *
 * @param argc Número de argumentos
 * @param argv Argumentos de línea de comandos
 * @return 0 si ejecuta correctamente
 */
int main(int argc, char* argv[]) {
    // -- Valores por defecto --
    int numPedidos = 100;
    double dMax = 0.015;
    int capMax = 15;

    // -- Parsear argumentos de línea de comandos --
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--pedidos" && i + 1 < argc) {
            numPedidos = std::stoi(argv[++i]);
        } else if (arg == "--dmax" && i + 1 < argc) {
            dMax = std::stod(argv[++i]);
        } else if (arg == "--capmax" && i + 1 < argc) {
            capMax = std::stoi(argv[++i]);
        }
    }

    // -- Encabezado --
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "  GEOCERCAS DINAMICAS - Clave del Peluquero Exitoso (C++)" << std::endl;
    std::cout << "  Centro: Caldas, Antioquia (" << CENTRO_LAT << ", " << CENTRO_LNG << ")" << std::endl;
    std::cout << std::string(60, '=') << "\n" << std::endl;

    // -- Generar pedidos --
    std::cout << "Generando " << numPedidos << " pedidos aleatorios..." << std::endl;
    auto pedidos = generarPedidos(numPedidos);

    // -- Ejecutar algoritmo con medición de tiempo --
    std::cout << "Ejecutando algoritmo (dMax=" << dMax
              << ", capMax=" << capMax << ")..." << std::endl;

    auto inicio = std::chrono::high_resolution_clock::now();

    std::vector<Centroide> centroides;
    auto grupos = agruparGeocercas(pedidos, dMax, capMax, centroides);

    auto fin = std::chrono::high_resolution_clock::now();

    // -- Calcular métricas --
    double tiempo_ms = std::chrono::duration<double, std::milli>(fin - inicio).count();
    // Estimación de memoria: sizeof(Pedido) * N + overhead
    double memoria_mb = (sizeof(Pedido) * numPedidos) / (1024.0 * 1024.0);

    // -- Mostrar resultados --
    std::cout << "\n" << std::string(40, '-') << std::endl;
    std::cout << "  RESULTADOS" << std::endl;
    std::cout << std::string(40, '-') << std::endl;
    std::cout << "  Lenguaje:        C++" << std::endl;
    std::cout << "  Pedidos:         " << numPedidos << std::endl;
    std::cout << "  Geocercas:       " << grupos.size() << std::endl;
    std::cout << std::fixed << std::setprecision(3);
    std::cout << "  Tiempo:          " << tiempo_ms << " ms" << std::endl;
    std::cout << "  Memoria (est):   " << memoria_mb << " MB" << std::endl;
    std::cout << std::string(40, '-') << "\n" << std::endl;

    // -- Detalle por geocerca --
    for (size_t i = 0; i < grupos.size(); ++i) {
        double pesoTotal = 0;
        for (const auto& p : grupos[i]) pesoTotal += p.peso;

        std::cout << "  Geocerca #" << (i + 1) << ": "
                  << grupos[i].size() << " pedidos, "
                  << std::setprecision(2) << pesoTotal << " kg, "
                  << "centro (" << std::setprecision(4)
                  << centroides[i].lat << ", " << centroides[i].lng << ")"
                  << std::endl;
    }

    std::cout << "\nAlgoritmo completado exitosamente.\n" << std::endl;

    return 0;
}

// ==========================================================================
// FIN DEL ARCHIVO GEOCERCAS.CPP
// ==========================================================================
